### ISSUE TITLE
Support multi-GPU training and prediction

### ISSUE THREAD

**Issue by AdeelH** [azavea/raster-vision#1352]
## 🚀 Feature
Support multi-GPU training and prediction.

## Motivation

Distributing the workload across multiple GPUs can considerably speed up the training and prediction process..

## Pitch

PyTorch provides [at least two ways](https://pytorch.org/tutorials/beginner/dist_overview.html) of utilizing multiple GPUs:
- `DataParallel`
- `DistributedDataParallel`

`DataParallel` is trivial to implement but is not very efficient. `DistributedDataParallel` is more efficient and is the recommended approach, but more complicated to implement. I think Raster Vision should support both.

### `DataParallel`
Using `DataParallel` would be as simple as wrapping the model at the end of `Learner.setup_model()`:
```python
self.model = DataParallel(self.model)
```
Plus small changes to ensure saving/loading weights works fine.

### `DistributedDataParallel`
Use of `DistributedDataParallel` involves spawning multiple processes and initializing the model, loss function, and optimizer within each process.

This could be done by moving `Learner.setup_training()` inside the function that each process runs after the fork. Doing this for prediction might be trickier and might require modifying the learner backend.

Saving/loading weights and logging to TensorBoard will also tricky.

## Alternatives

N/A

## Additional context

Resources:
- https://pytorch.org/tutorials/beginner/dist_overview.html
- https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
- https://pytorch.org/docs/stable/notes/multiprocessing.html
- https://medium.com/codex/a-comprehensive-tutorial-to-pytorch-distributeddataparallel-1f4b42bb1b51 (published Aug, 2021 -- might be outdated)

**Comment by julled** [comment]
I am very much in need for this feature

**Comment by AdeelH** [comment]
An alternative to using Raster Vision for training is to use PyTorch Lightning, which, I believe, does support multi-GPU training. You can still use Raster Vision for reading data.

Here is an example: https://docs.rastervision.io/en/latest/usage/tutorials/lightning_workflow.html

**Comment by JamesZDonline** [comment]
It looks like the new library (instead of framework) approach opens up a lot more of these kinds of integrations, which is great. Do you have any ideas about how to get multi-GPU to work on the prediction end? We're working on some areas at the scale of 300k+ sqkm and it would be great to speed up the predictions.

**Comment by AdeelH** [comment]
>Do you have any ideas about how to get multi-GPU to work on the prediction end?

I found this discussion on using multi-GPU prediction using lightning: https://github.com/Lightning-AI/lightning/discussions/9259.

CC'ing @lewfish, in case he has any info on this.

**Comment by lewfish** [comment]
I think the simplest way to make predictions in parallel would be to cut your image into smaller pieces so that multiple instances of the `predict` command can run in parallel over them. 

If you are using Lightning, you could override the [`BasePredictionWriter`](https://pytorch-lightning.readthedocs.io/en/stable/api/pytorch_lightning.callbacks.BasePredictionWriter.html) class which make predictions over the items in a `Dataset` in parallel, and you can implement custom logic for what to do with each prediction. In our case the items would be chips within a scene, and you would probably want to append the predictions to a `Labels` object.

**Comment by JamesZDonline** [comment]
@lewfish Splitting the data up makes good sense, and I'm a bit ashamed not to have thought of it myself. It's easy to get myopic about a particular solution sometimes.

Thank you both for the help!
